lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -802,6 +803,263 @@ func::FuncOp PolynomialToStandard::buildPolynomialModFunc(FunctionType funcType,
   return funcOp;
 }
 
+// Multiply two integers x, y modulo cmod.
+static APInt mulMod(const APInt &_x, const APInt &_y, const APInt &_cmod) {
+  assert(_x.getBitWidth() == _y.getBitWidth() &&
+         "expected same bitwidth of operands");
+  auto intermediateBitwidth = _cmod.getBitWidth() * 2;
+  APInt x = _x.zext(intermediateBitwidth);
+  APInt y = _y.zext(intermediateBitwidth);
+  APInt cmod = _cmod.zext(intermediateBitwidth);
+  APInt res = (x * y).urem(cmod);
+  return res.trunc(_x.getBitWidth());
+}
+
+// Compute the first degree powers of root modulo cmod.
+static SmallVector<APInt> precomputeRoots(APInt root, const APInt &cmod,
+                                          unsigned degree) {
+  APInt baseRoot = root;
+  root = 1;
+  SmallVector<APInt> vals(degree);
+  for (unsigned i = 0; i < degree; i++) {
+    vals[i] = root;
+    root = mulMod(root, baseRoot, cmod);
+  }
+  return vals;
+}
+
+static Value computeReverseBitOrder(ImplicitLocOpBuilder &b,
+                                    RankedTensorType type, Value tensor) {
+  unsigned degree = type.getShape()[0];
+  double degreeLog = std::log2((double)degree);
+  assert(std::floor(degreeLog) == degreeLog &&
+         "expected the degree to be a power of 2");
+
+  unsigned indexBitWidth = (unsigned)degreeLog;
+  auto indicesType =
+      RankedTensorType::get(type.getShape(), IndexType::get(b.getContext()));
+
+  SmallVector<APInt> _indices(degree);
+  for (unsigned index = 0; index < degree; index++) {
+    _indices[index] = APInt(indexBitWidth, index).reverseBits();
+  }
+  auto indices = b.create<arith::ConstantOp>(
+      indicesType, DenseElementsAttr::get(indicesType, _indices));
+
+  SmallVector<utils::IteratorType> iteratorTypes(1,
+                                                 utils::IteratorType::parallel);
+  AffineExpr d0;
+  bindDims(b.getContext(), d0);
+  SmallVector<AffineMap> indexingMaps = {AffineMap::get(1, 0, {d0}),
+                                         AffineMap::get(1, 0, {d0})};
+  auto out = b.create<arith::ConstantOp>(
+      type,
+      DenseElementsAttr::get(type, APInt(type.getElementTypeBitWidth(), 0L)));
+  auto shuffleOp = b.create<linalg::GenericOp>(
```

Comment:
I don't have a ton of experience with this, but I _think_ you should be able to replace this with a `tensor.gather` op, using the indices created above as the index argument.

Feel free to leave as-is, but using a dedicated op could make this code simpler.

Reply:
Unfortunately when trying to use `gather` it requires the output tensor to be strictly a subset of the source.